gen.go

```diff
@@ -169,10 +173,84 @@ func typeName(pkg string, t reflect.Type) string {
 			// It's a built-in.
 			return t.String()
 		} else if pkgPath == pkg {
-			return t.Name()
+			return genericName(t.Name())
+		}
+		return fmt.Sprintf("%s.%s", resolvePkgName(pkgPath, t.String()), genericName(t.Name()))
+	}
+}
+
+func getGenericTypeParameter(index int) string {
+	return "T" + strconv.Itoa(index+1)
+}
+
+// nameForms returns the type name without generic specifier, a generated name
+// for use in variable names, and a list of generic placeholders if the type is
+// a generic type.
+func nameForms(typeName string) (string, string, []string) {
+	re := regexp.MustCompile(`^(.*?)\[(.*)\]$`)
+	matches := re.FindStringSubmatch(typeName)
+
+	name := typeName
+	generatedName := typeName
+	var genericPlaceholders []string
+
+	if len(matches) == 3 {
+		generatedName = matches[1]
+		genericPlaceholders = strings.Split(matches[2], ",")
+		for i := range genericPlaceholders {
+			genericPlaceholders[i] = strings.TrimSpace(genericPlaceholders[i])
+		}
+		var genericTypes []string
+		for i := range genericPlaceholders {
+			genericTypes = append(genericTypes, getGenericTypeParameter(i))
+		}
+		name = matches[1] + "[" + strings.Join(genericTypes, ",") + "]"
+	}
+
+	return name, generatedName, genericPlaceholders
+}
+
+func genericName(name string) string {
+	name, _, _ = nameForms(name)
+	return name
+}
+
+// toGeneratedName replaces [GenericPlaceholder] with [T] for nicer comment printing
+func (gti GenTypeInfo) toGeneratedName(name string) string {
+	// if this is a placeholder, replace it with the generic type param
+	for i, placeholder := range gti.GenericPlaceholders {
+		genericType := getGenericTypeParameter(i)
+		trimmedPlaceholder := strings.TrimPrefix(placeholder, "*")
+		if name == placeholder || name == trimmedPlaceholder {
+			return genericType
+		}
+	}
+
+	// if this is a type that has generics, replace the placeholders with the corresponding generic type params
+	re := regexp.MustCompile(`\[(.*?)\]`)
+	return re.ReplaceAllStringFunc(name, func(match string) string {
+		inner := match[1 : len(match)-1]
+		parts := strings.Split(inner, ",")
+		for j, part := range parts {
+			for i, placeholder := range gti.GenericPlaceholders {
+				if part == placeholder {
+					parts[j] = getGenericTypeParameter(i)
+				}
+			}
+		}
+		return "[" + strings.Join(parts, ",") + "]"
+	})
+}
+
+func (gti GenTypeInfo) toGenericTypeParam(name string) string {
+	name = strings.TrimPrefix(name, "*")
```

Comment:
left this comment:

```
	// dummy generic placeholders should be non-pointers at generation time, but they may be used
	// as pointers (e.g. struct `Foo[T]` but Foo has a field `F *T``) so we need to strip the
	// pointer prefix if there is one to match the placeholder
```

This shows up in the tests:

1. You can't supply a pointer generic, `Foo[*dummy]` just doesn't work because codegen can't deal with pointer dummy types, if you want to use generic `T` as a `*T` somewhere in the code then you're in trouble:

```
panic: failed to generate encoders: field "t.Sub2" of "SubGenericStruct[T1,T2]" has unsupported kind "ptr"
```

So at generate time, dummy types are always values and very simple, but you can use them as `T` or `*T` just fine, so we end up matching `dummy` and `*dummy` against what we have a record of: `dummy`.

But, if you never use `*T` then you can technically still use a `*dummy` as a pointer even though you probably shouldn't, so we add the second `*` prefix trim just to be sure we're matching the right thing.

Reply:
forgot to push changes, done now